internal/pkg/client/client.go

```diff
@@ -47,3 +55,85 @@ type (
 		Push(pushURL string, opts *options.PushOptions) error
 	}
 )
+
+// URL: [scheme:][//[userinfo@]host][/]path[?query][#fragment]
+//
+// https://username:password@github.com:12345/bomctl/bomctl.git?ref=main#sbom.cdx.json
+//
+// scheme: https
+// userinfo: [username, password]
+// host: github.com:12345 (parse port with url.Port())
+// path: bomctl/bomctl.git
+// fragment: sbom.cdx.json
+// Query: ref=main (parse query with url.Query() and get value with query.Get("ref"))
+//
+
+func NewFetcher(fetchURL string) (Fetcher, error) {
+	var fetch Fetcher
+
+	c, err := New(fetchURL)
+	fetch, ok := c.(Fetcher)
+
+	if err != nil || !ok {
+		return nil, ErrUnsupportedURL
+	}
+
+	return fetch, nil
+}
+
+func New(sbomURL string) (Client, error) {
+	parsedURL, err := neturl.Parse(sbomURL)
+	if err != nil {
+		return nil, ErrUnsupportedURL
+	}
+
+	return DetermineClient(parsedURL)
+}
+
+func DetermineClient(parsedURL *neturl.URL) (Client, error) { //nolint:cyclop
+	if client, err := checkScheme(parsedURL); err == nil {
+		return client, nil
+	}
+
+	switch {
+	case strings.Contains(parsedURL.Path, ".git") || parsedURL.Fragment != "":
+		return &git.Client{}, nil
+	case strings.Contains(parsedURL.Host, "github") && !strings.Contains(parsedURL.Host, "githubusercontent"):
+		return &github.Client{}, nil
+	case strings.Contains(parsedURL.Host, "gitlab"):
+		return &gitlab.Client{}, nil
+	case sliceutil.Any(registrySlice(), func(s string) bool { return strings.Contains(parsedURL.Host, s) }):
+		//nolint:wrapcheck
+		return oci.Init(parsedURL)
+	case parsedURL.Scheme == "https" || parsedURL.Scheme == "http":
+		return &http.Client{}, nil
+	default:
+		return nil, fmt.Errorf("%w", errors.ErrUnsupported)
+	}
+}
+
+func checkScheme(parsedURL *neturl.URL) (Client, error) {
+	if parsedURL.Scheme == "" {
+		parsedURL.Scheme = "https"
+	}
+
+	// check for explicit selection via scheme
+	switch parsedURL.Scheme {
+	case "git":
+		return &git.Client{}, nil
+	case "github":
+		return &github.Client{}, nil
+	case "gitlab":
+		return &gitlab.Client{}, nil
+	case "oci", "oci-archive", "docker", "docker-archive":
+		//nolint:wrapcheck
+		return oci.Init(parsedURL)
+	default:
+		return nil, fmt.Errorf("%w", errors.ErrUnsupported)
+	}
+}
+
+func registrySlice() []string {
+	// maybe check for env variable like BOMCTL_REGISTRY or something and append to slice if it exists
+	return []string{"registry", "harbor", "ghcr.io", "gcr.io", "docker.io"}
```

Comment:
```suggestion
	return []string{"quay.io", "ghcr.io", "gcr.io", "docker.io"}
```
Remove items that are too general and likely to cause collision/mistaken match, add only known registries